python/cudf/cudf/core/column/categorical.py

```diff
@@ -967,26 +973,42 @@ def find_and_replace(self, to_replace, replacement, all_nan):
         """
         Return col with *to_replace* replaced with *replacement*.
         """
-        replaced = column.as_column(self.cat().codes)
 
-        to_replace_col = column.as_column(
-            np.asarray(
-                [self._encode(val) for val in to_replace], dtype=replaced.dtype
-            )
+        old_cats = cudf.DataFrame()
+        old_cats["cats"] = column.as_column(self.dtype.categories)
+        new_cats = old_cats.copy(deep=True)
+
+        old_cats["cats_replace"] = old_cats["cats"].replace(
+            to_replace, replacement
         )
-        replacement_col = column.as_column(
-            np.asarray(
-                [self._encode(val) for val in replacement],
-                dtype=replaced.dtype,
-            )
+        
+        dtype_replace = cudf.Series(replacement)
+        dtype_replace[dtype_replace.isin(old_cats["cats"])] = None
+        new_cats["cats"] = new_cats["cats"].replace(to_replace, dtype_replace)
+
+        new_cats = (
+            new_cats[new_cats["cats"].notna()]
```

Comment:
This will boolean mask the index as well unnecessarily. Could we do this on the column instead?

Reply:
We actually want to gather the index here. The index gives us the set of original integers we need to map to new ones in the underlying int column. This is to avoid having gaps in the underlying integers that represent our categories.

Reply:
You drop it immediately after this function though

Reply:
You're right - got mixed up here. updated